### ISSUE TITLE
Thoughts on how `overrides` works

### ISSUE THREAD

**Issue by DeyLak** [teambit/bit#2014]
### Description

This is not exactly a feature request, but as @itaymendel asked in Gitter, I'm trying to formalise my thoughts on how `overrides` config should work.

Now `overrides` is working like this:
1. You import some component to your workspace
2. `overrides` adds it's keys to imported component, but doesn't exactly override them. For example, `compiler` option is only applied, if a component doesn't have a compiler.
3. Bit detects change in the component and requires to tag the "overridden" keys

I see some problems with this approach:
1. Impossible to have different `overrides` config per project (components will always needed to be tagged after import between projects)
2. `overrides` config becomes unintuitive and doesn't work, when it doesn't _override_ component's `package.json` keys. 
3. `overrides` config doesn't have big value, cause I can make the same changes directly to . component's `package.json` (I will need to tag it anyway)

### Describe the solution you'd like

I'd like to think about `overrides` as a project-level only instrument, which is not influencing the collection, I'm using. I think, that `overrides` should prevail component's `package.json`, but it should not be considered a change in the component. That makes this config a tuning instrument, similar to Bit capability of having a non-tagged project-level changes in the component. I believe, such opportunities are really cool features, which are not only smoothing Bit integration with existing codebases, but also may helping reducing complexness in mature component libraries, that are used in large number of projects, by moving some specific functionality back to the project-level.

**Comment by itaymendel** [comment]
Before I dive into explaining some of the thought process behind `overrides`, I want to say that you are free to modify a component's dependency graph (with `overrides`, or any other way) as much as you'd like, in each workspace. You can either use the per-component configs, or have a global configuration that can control entire sub-sections of your component tree.

#### Ignoring overrides

A component is not just a few files with code written in them. A major part of its functionality is its dependencies, extensions and configuration. This is why, as part of the `bit tag` process, modifications to the dependency graph are logged as well. 
The graph is important to the component's functionality and for the reproducibility of the entire isolated component environment in each consuming project.

`overrides` is a component-level tool that designed to allow developers a way to interact with the automatically generated dependency graph (or any other config for that matter). Diverging it from this behavior might cause situations where a component that worked in one workspace will not work in another, due to missing dependencies (or mixed-versions).

When you want to version a component, I do think that it is important to keep the full state of it in the version, including all local overrides to the dependency graph. The same thing goes with extensions such as build or test environment.

#### Overriding envs

From your note, it seems that for some reason Bit failed to -override- a component compiler. This seems to be more of a bug rather than an actual implemented behavior. I'd like to look into this issue regardless of this discussion.

#### Additional things to consider

I do see the value of adding a capability to *ignore* `overrides` for imported components. This is primerily due to the inherited difference with the way that you build and interact with components that are originated in a repo and remained as a part of it (ie - components that where never ejected).

**Comment by GiladShoham** [comment]
@DeyLak 
Sometimes, for example, you are importing a lot of components and want to change something for all of them at once, for example, the compiler version. or a dependency version. (A change that you want to tag and export for all of them) in such cases, it will be much more convenient to do it with a glob pattern in the root's config. 
We do understand the confusion and we are in the process of making it more clear and intuitive.

**Comment by DeyLak** [comment]
@GiladShoham I see the convenience in using `overrides` with glob patterns. On the other hand, I don't see, why should it be stored in Bit's config. It causes duplication in components configs and global configuration. Other collaborator may have different `overrides` in his workspace for different purposes, so he will have to change it to match new imported components. My initial thought about project-level configuration was based on this problem. 
I don't see much difference between running something like `bit override "types/" "{ "env": { "compiler": "-" }}"` and having the same in `bit.json`, cause this change is applied only once and doesn't have to be stored and committed into VCS. Ofcourse, json configs are not easy to type in console, but there could be some args based interface like `bit override "types/" --compiler "-" --devDependencies "@types/node@^11.13.4"`, or with config file like: `bit override --conf etc/override.json`. 
To summarise, I think that, 'overrides' config is made to feel declarative, but it's really not, This is one of the sources of confusion.

For authored components, as far as I can see, `overrides` config is the only place, where user can change compilers, deps etc. Should it be considered as different problem? Maybe they should have theirs `package.json` and `bit.json` files somewhere to make the api more universal?

**Comment by GiladShoham** [comment]
Thanks for the feedback, we will process it and come up with a better product hopefully.

Regarding authored components - 
The source of the issue is that as opposed to imported components, an authored component might be in one of the following structures:
1. Different files of the same component are in different paths, for example:
```
src/components/my-button.js
test/components/my-button.test.js
```
2. more than one component in the same dir, for example:
```
src/utils/sort.js (a component)
src/utils/group.js (another component)
```
These 2 cases prevent from writing a package.json for the authored component, since you don't know where to put them in the first case, and conflicts between components in the second case.

I develop a feature which enables you to run `bit eject-config` command and let you specify a file where you want to store a component configuration. This file location and name are defined by the user. so both problems solved.
This feature is pretty old, but we never documented it since it was never well tested and stabilized. 
It's been written a lot before the override feature, and we thought it's pretty much redundant now that we have the override.
According to how you see it, it's still necessary, since they are a bit different cases between project overrides and component config files.
We will consider continue working on this and stabilize it according to your inputs.

**Comment by itaymendel** [comment]
`overrides` is going to be replaces by `variants` with v15 of Bit later this week that solves these feedbacks.

We'll start rolling out v15 of Bit this week, please reach out to me privately on the [public slack community](https://join.slack.com/t/bit-dev-community/shared_invite/zt-9egg85xv-trr_QBkWr1XcVgtrgCvtfQ) for a sneak peek and getting early access to it.
